### PR TITLE
Add taint-preserving edges where a call also has a value-preserving edge

### PR DESCRIPTION
**Pull request by smowton**
For example, for a fluent method that returns `this`, we take a tainting edge from argX to either `this` or the return value to also taint the other.

### CODE REVIEW

File: java/ql/src/semmle/code/java/dataflow/internal/TaintTrackingUtil.qll [hunk 1]
```diff
@@ -56,6 +63,31 @@ predicate localAdditionalTaintStep(DataFlow::Node src, DataFlow::Node sink) {
...
 
+/**
+ * Holds if an additional step from `src` to `sink` can be inferred from a value-preserving step
+ * across a method callsite (from input to input, or input to result) and a taint-preserving step
+ * across the same callsite from a different input. For example, if we know that `f(a, b)` returns
+ * the exact value of `a` and also propagates taint from `b` to its result, then we also know that
+ * `a` is tainted after `f` completes, and vice versa.
+ */
+private predicate composedValueAndTaintModelStep(ArgumentNode src, DataFlow::Node sink) {
+  exists(Call call, ArgumentNode valueSource, DataFlow::PostUpdateNode valueSourcePun |
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/dataflow/internal/TaintTrackingUtil.qll:74]:
```suggestion
  exists(Call call, ArgumentNode valueSource, DataFlow::PostUpdateNode valueSourcePost |
```

Comment (aschackmull) [reply java/ql/src/semmle/code/java/dataflow/internal/TaintTrackingUtil.qll:74]: (I think that's a slightly better name)

File: java/ql/src/semmle/code/java/dataflow/internal/TaintTrackingUtil.qll [hunk 2]
```diff
@@ -56,6 +63,31 @@ predicate localAdditionalTaintStep(DataFlow::Node src, DataFlow::Node sink) {
...
+    src.argumentOf(call, _) and
+    valueSource.argumentOf(call, _) and
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/dataflow/internal/TaintTrackingUtil.qll:76]: Probable add `src != valueSource` for good measure.

**Comment by aschackmull** [comment]
Two minor suggestions, otherwise LGTM.

File: java/ql/src/semmle/code/java/dataflow/internal/TaintTrackingUtil.qll [hunk 3]
```diff
@@ -56,6 +63,31 @@ predicate localAdditionalTaintStep(DataFlow::Node src, DataFlow::Node sink) {
   )
 }
 
+/**
+ * Holds if an additional step from `src` to `sink` can be inferred from a value-preserving step
+ * across a method callsite (from input to input, or input to result) and a taint-preserving step
+ * across the same callsite from a different input. For example, if we know that `f(a, b)` returns
+ * the exact value of `a` and also propagates taint from `b` to its result, then we also know that
+ * `a` is tainted after `f` completes, and vice versa.
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/dataflow/internal/TaintTrackingUtil.qll:71]: We can't have a value-step from input to input. Maybe rephrase to something like this:
```suggestion
 * Holds if an additional step from `src` to `sink` through a call can be inferred from the
 * combination of a value-preserving step providing an alias between an input and the output
 * and a taint step from `src` to one the aliased nodes. For example, if we know that `f(a, b)` returns
 * the exact value of `a` and also propagates taint from `b` to its result, then we also know that
 * `a` is tainted after `f` completes, and vice versa.
```

**Comment by smowton** [comment]
@aschackmull done

**Review by aschackmull — APPROVED** [review]
